icu4c/source/common/uloc.cpp

```diff
@@ -1184,104 +1184,128 @@ _getLanguage(const char* localeID,
         localeID += 3;
     }
 
+    constexpr int32_t MAXLEN = ULOC_LANG_CAPACITY - 1;  // Minus NUL.
+
     /* if it starts with i- or x- then copy that prefix */
-    if(_isIDPrefix(localeID)) {
-        result.append((char)uprv_tolower(*localeID), status);
-        result.append('-', status);
-        localeID+=2;
+    int32_t len = _isIDPrefix(localeID) ? 2 : 0;
+    while (!_isTerminator(localeID[len]) && !_isIDSeparator(localeID[len])) {
+        if (len == MAXLEN) {
```

Comment:
optional / not a new problem in this PR: we apparently allow a length of 1, and allow characters other than letters...